Fix #148: Update the tab bar's visibility based on the preference

kylehickinson opened the pull request:
## Pull Request Checklist

- [ ] My patch has gone through review and I have addressed review comments
- [x] My patch has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!`

- [ ] I have updated the *Unit Tests* to cover new or changed functionality
- [ ] I have updated the *UI Tests* to cover new or changed functionality
- [ ] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

_none included_

## Notes for testing this patch

_none included_

Commit:
Fix #148: Update the tab bar's visibility based on the preference

iccub commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2606,7 +2623,7 @@ extension BrowserViewController: TopSitesDelegate {
     func preferencesDidChange(for key: String) {
         HTTPCookieStorage.shared.updateCookieAcceptPolicy(to: HTTPCookie.AcceptPolicy(rawValue: Preferences.Privacy.cookieAcceptPolicy.value))
-        // TODO: Update tab bar visiblity based on `Preferences.tabBarVisibility` once BraveURLBarView is back
+        updateTabsBarVisibility()
```
Shouldn't we utilize `key` param here to update tab bar visibility only if tabs bar related preferences are changed? Now it triggers on every preference change. Same goes with HTTP cookie storage line above.

Now every shields popup settings tick would call all this code, maybe I'm missing something.

kylehickinson replied:
Good call, made me realize I didn't even setup observing for the `Preferences.General.tabBarVisibility` preference at all.

Updated so it only does relevant actions based on key

iccub commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2605,12 +2626,21 @@ extension BrowserViewController: TopSitesDelegate {
+        case Preferences.Shields.fingerprintingProtection.key:
+            // TODO: Update fingerprinting protection based on `Preferences.Shields.fingerprintingProtection` once fingerprinting protection is added back
+            break
+        default:
```
Add a `log` for default case and/or assertion if that should never happen.

kylehickinson replied:
Added a log

iccub reviewed (changes requested):
Just add a log and rest looks good.

Small inconvenience is needing to add observer for each property, no catch-all option like with `NSNotification` for example.

jhreis commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -750,8 +754,25 @@ class BrowserViewController: UIViewController {
     
     func updateTabsBarVisibility() {
+        func shouldShowTabBar() -> Bool {
+            let tabCount = UIApplication.isInPrivateMode ? tabManager.privateTabs.count : tabManager.normalTabs.count
```
We should remove the concept of `privateTabs` and `normalTabs` to just `tabs`. Let the result be returned solely based on the current mode. This may be too large of a refactor for now, since it completely breaks the TabManager's concept of private tab (e.g. no longer utilizing coredata).

jhreis replied:
Also I think using `isEmpty` is probably better than using all of the `> 1` checks later down. Just make `tabCount` a boolean `hasTabs` or something.

kylehickinson replied:
I can dump the use if private vs normal tabs, but isEmpty doesn't check for > 1, it checks for > 0, so it would show the tab bar even if there's only 1 tab. If thats cool, I'll do that.

kylehickinson committed:
Ignore private vs normal tabs for updating tab bar visibility

Also remove the delay on updating the tab bar which was causing some UI issues when only 1 tab remained
